README.md
# NAME

Google::CloudTasks - Perl client library for the Google CloudTasks API (_unofficial_).

# SYNOPSIS

    use Google::CloudTasks;

    my $client = Google::CloudTasks->client(
        version => 'v2',
        credentials_path => '/path/to/credentials.json',
    );

    #  Create task
    my $project_id = 'myproject';
    my $location_id = 'asia-northeast1';
    my $queue_id = 'myqueue';
    my $parent = "projects/$project_id/locations/$location_id/queues/$queue_id";

    my $task = {
        name => "$parent/tasks/mytask-01234567",
        appEngineHttpRequest => {
            relativeUri => '/do_task',
        },
    }
    my $ret = $client->create_task($parent, $task);

# DESCRIPTION

Google::CloudTasks [https://cloud.google.com/tasks/docs/reference/rest/](https://cloud.google.com/tasks/docs/reference/rest/)

This is a Perl client library for the Google CloudTasks API.

## AUTHENTICATION

A service account with appropriate roles is required. You need to download JSON file and specify `credentials_path`.
See also: [https://cloud.google.com/docs/authentication/getting-started#creating\_the\_service\_account](https://cloud.google.com/docs/authentication/getting-started#creating_the_service_account)

# METHODS

All methods handle raw hashref (or arrayref of hashref), rather than objects.

## Create a client

    my $client = Google::CloudTasks->client(
        version => 'v2',
        credentials_path => '/path/to/credentials.json',
    );

`version` is an API version. (Currently only `v2` is available)
`credentials_path` is a path to a service account JSON file.

## Location

Refer the detailed representation of location at [https://cloud.google.com/tasks/docs/reference/rest/Shared.Types/ListLocationsResponse#Location](https://cloud.google.com/tasks/docs/reference/rest/Shared.Types/ListLocationsResponse#Location)

### get\_location

Gets information about a location.

    my $location = $client->get_location("projects/$PROJECT_ID/locations/$LOCATION_ID");

### list\_locations

Lists information about all locations under project.

    my $ret = $client->list_locations("projects/$PROJECT_ID");
    my $locations = $ret->{locations};

## Queue

Refer the detailed representation of queue at [https://cloud.google.com/tasks/docs/reference/rest/v2/projects.locations.queues#Queue](https://cloud.google.com/tasks/docs/reference/rest/v2/projects.locations.queues#Queue)

### create\_queue

Creates a queue.

    my $queue = {
        name => 'queue-name',
    };
    my $created = $client->create_queue("projects/$PROJECT_ID/locations/$LOCATION_ID", $queue);

### delete\_queue

Deletes a queue.

    $client->delete_queue("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID")

### get\_queue

Gets information of a queue.

    my $queue = $client->get_queue("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID");

### list\_queues

Lists information of all queues.

    my $ret = $client->list_queues("projects/$PROJECT_ID/locations/$LOCATION_ID");
    my $queues = $ret->{queues};

### patch\_queue

Updates a queue.

    my $queue = {
        retryConfig => {
            maxAttempts => 5,
        },
    };
    my $update_mask = { updateMask => 'retryConfig.maxAttempts' };
    my $updated = $client->patch_queue(
        "projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID",
        $queue,
        $update_mask,   # optional
    );

### pause\_queue

Pauses a queue.

    my $queue = $client->pause_queue("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID");

### resume\_queue

Resumes a queue.

    my $queue = $client->resume_queue("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID");

## Task

Refer the detailed representation of task at [https://cloud.google.com/tasks/docs/reference/rest/v2/projects.locations.queues.tasks#Task](https://cloud.google.com/tasks/docs/reference/rest/v2/projects.locations.queues.tasks#Task)

### create\_task

Creates a task. Note that a request body in `appEngineHttpRequest` should be base64-encoded.

    use MIME::Base64;

    my $body = encode_base64('{"name": "TaskTest"}');
    chomp($body);

    my $task = {
        name => "projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID",
        appEngineHttpRequest => {
            relativeUri => '/path',
            headers => [
                'Content-Type' => 'application/json',
            ],
            body => $body,
        },
    };
    my $created = $client->create_task(
        "projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID",
        $task
    );

### delete\_task

Deletes a task.

    $client->delete_task("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID/tasks/$TASK_ID");

### get\_task

Gets information of a task.

    my $task = $client->get_task("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID/tasks/$TASK_ID");

### list\_tasks

Lists information of all tasks.

    my $ret = $client->list_tasks("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID");
    my $tasks = $ret->{tasks};

### run\_task

Runs a task.

    my $ret = $client->run_task("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID/tasks/$TASK_ID");

# TODO

The following methods has implemented, but not tested nor documented yet.

`Queue.getIamPolicy`

`Queue.setIamPolicy`

`Queue.testIamPermissions`

# LICENSE

Copyright (C) egawata.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

egawata (egawa dot takashi at gmail.com)


Changes
Revision history for Perl extension Google-CloudTasks

{{$NEXT}}

0.01 2019-08-12T05:35:05Z

    - original version



lib/Google/CloudTasks/Queue.pm
package Google::CloudTasks::Queue;

our $VERSION = "0.01";

use Mouse;

has client => (
    is => 'ro',
    isa => 'Google::CloudTasks::Client',
    required => 1,
    weak_ref => 1,
);

has name => (
    is => 'rw',
    isa => 'Str',
);

has rateLimits => (
    is => 'rw',
    isa => 'Google::CloudTasks::RateLimits',
);

has retryConfig => (
    is => 'rw',
    isa => 'Google::CloudTasks::RetryConfig',
);

has state => (
    is => 'rw',
    isa => 'Str',
);

has purgeTime => (
    is => 'rw',
    isa => 'Str',
);

has appEngineRouting => (
    is => 'rw',
    isa => 'Google::CloudTasks::AppEngineRouting',
);

no Mouse;

__PACKAGE__->meta->make_immutable;

sub new_from_hash {
    my ($class, $client, $hash) = @_;

    my $obj = __PACKAGE__->new(
        client => $client,
    );

    for (qw/name state purgeTime/) {
        if (defined $hash->{$_}) {
            $obj->$_($hash->{$_});
        }
    }

    if ($hash->{rateLimits}) {
        $obj->rateLimits(Google::CloudTasks::RateLimits->new($hash->{rateLimits}));
    }
    if ($hash->{retryConfig}) {
        $obj->retryConfig(Google::CloudTasks::RetryConfig->new($hash->{retryConfig}));
    }

    return $obj;
}

sub delete {
    my ($self) = @_;

    return $self->client->request_delete($self->name);
}

package Google::CloudTasks::RateLimits;

use Mouse;

has maxDispatchesPerSecond => (is => 'rw', isa => 'Num');
has maxBurstSize => (is => 'rw', isa => 'Num');
has maxConcurrentDispatches => (is => 'rw', isa => 'Num');

no Mouse;

__PACKAGE__->meta->make_immutable;

package Google::CloudTasks::RetryConfig;

use Mouse;

has maxAttempts => (is => 'rw', isa => 'Num');
has maxRetryDuration => (is => 'rw', isa => 'Str');
has minBackoff => (is => 'rw', isa => 'Str');
has maxBackoff => (is => 'rw', isa => 'Str');
has maxDoublings => (is => 'rw', isa => 'Num');

no Mouse;

__PACKAGE__->meta->make_immutable;

package Google::CloudTasks::AppEngineRouting;

use Mouse;

has service => (is => 'rw', isa => 'Str');
has version => (is => 'rw', isa => 'Str');
has instance => (is => 'rw', isa => 'Str');
has host => (is => 'rw', isa => 'Str');

no Mouse;

__PACKAGE__->meta->make_immutable;

1;


lib/Google/CloudTasks/Client.pm
package Google::CloudTasks::Client;

use Mouse;
use WWW::Google::Cloud::Auth::ServiceAccount;
use LWP::UserAgent;
use HTTP::Request;
use URI;
use URI::QueryParam;
use JSON::XS;

our $VERSION = "0.01";

has base_url => (
    is => 'ro',
    isa => 'Str',
    default => 'https://cloudtasks.googleapis.com/',
);

has version => (
    is => 'ro',
    isa => 'Str',
    default => 'v2',
);

has credentials_path => (
    is => 'ro',
    isa => 'Str'
);

has auth => (
    is => 'ro',
    lazy_build => 1,
);

has ua => (
    is => 'ro',
    lazy => 1,
    default => sub { LWP::UserAgent->new() },
);

has is_debug => (
    is => 'ro',
    isa => 'Bool',
    default => 0,
);

no Mouse;

__PACKAGE__->meta->make_immutable;

sub _build_auth {
    my ($self) = @_;

    if (!$self->credentials_path) {
        die "attribute 'credentials_path' is required";
    }
    my $auth = WWW::Google::Cloud::Auth::ServiceAccount->new(
        credentials_path => $self->credentials_path,
    );
    return $auth;
}

sub request {
    my ($self, $method, $path, $content) = @_;

    my $url = $self->base_url . $self->version . '/' . $path;
    my $req = HTTP::Request->new($method, $url);
    $req->header('Content-Type' => 'application/json; charset=utf8');
    $req->header('Authorization' => 'Bearer ' . $self->auth->get_token);
    if ($content) {
        my $encoded_body = encode_json($content);
        $req->header('Content-Length' => length($encoded_body));
        $req->content($encoded_body);
    }

    if ($self->is_debug) {
        use Data::Dumper;
        print "Request : " . Dumper($req);
    }

    my $res = $self->ua->request($req);

    if ($res->is_success) {
        return decode_json($res->content);
    }
    else {
        die "Fail: " . $res->content;
    }
}

sub request_get {
    my ($self, $path) = @_;
    return $self->request(GET => $path);
}

sub request_post {
    my ($self, $path, $content) = @_;
    $content //= {};
    return $self->request(POST => $path, $content);
}

sub request_delete {
    my ($self, $path) = @_;
    return $self->request(DELETE => $path);
}

sub request_patch {
    my ($self, $path, $content) = @_;
    return $self->request(PATCH => $path, $content);
}

sub _make_query_param {
    my ($args, @keys) = @_;

    my $u = URI->new();
    for (@keys) {
        if (defined $args->{$_}) {
            $u->query_param($_ => $args->{$_});
        }
    }

    return $u->query ? '?' . $u->query : '';
}

sub get_location {
    my ($self, $name) = @_;
    my $path = $name;

    return $self->request_get($path);
}

sub list_locations {
    my ($self, $name, $opts) = @_;
    my $path = $name . '/locations';
    $path .= _make_query_param($opts, qw/filter pageSize pageToken/);

    return $self->request_get($path);
}

sub create_queue {
    my ($self, $parent, $queue) = @_;
    my $path = $parent . '/queues';

    return $self->request_post($path, $queue);
}

sub delete_queue {
    my ($self, $name) = @_;
    my $path = $name;

    return $self->request_delete($path);
}

sub get_iam_policy {
    my ($self, $resource) = @_;
    my $path = $resource . ':getIamPolicy';

    return $self->request_post($path);
}

sub set_iam_policy {
    my ($self, $resource, $policy) = @_;
    my $path = $resource . ':setIamPolicy';

    return $self->request_post($path, { policy => $policy });
}

sub list_queues {
    my ($self, $parent, $opts) = @_;
    my $path = $parent . '/queues';
    $path .= _make_query_param($opts, qw/filter pageSize pageToken/);

    return $self->request_get($path);
}

sub get_queue {
    my ($self, $name) = @_;
    my $path = $name;

    return $self->request_get($path);
}

sub patch_queue {
    my ($self, $name, $queue, $opts) = @_;
    my $path = $name;
    $path .= _make_query_param($opts, qw/updateMask/);

    return $self->request_patch($path, $queue);
}

sub pause_queue {
    my ($self, $name) = @_;
    my $path = $name . ':pause';
    return $self->request_post($path);
}

sub purge_queue {
    my ($self, $name) = @_;
    my $path = $name . ':purge';
    return $self->request_post($path);
}

sub resume_queue {
    my ($self, $name) = @_;
    my $path = $name . ':resume';
    return $self->request_post($path);
}

sub test_iam_permissions {
    my ($self, $resource, $permissions) = @_;
    my $path = $resource . ':testIamPermissions';
    return $self->request_post($path, { permissions => $permissions });
}

sub create_task {
    my ($self, $parent, $task, $opts) = @_;
    my $path = $parent . '/tasks';

    my %param = (
        task => $task,
    );
    defined $opts->{responseView} and $param{responseView} = $opts->{responseView};

    return $self->request_post($path, \%param);
}

sub delete_task {
    my ($self, $name) = @_;
    my $path = $name;

    return $self->request_delete($path);
}

sub get_task {
    my ($self, $name, $opts) = @_;
    my $path = $name;

    $path .= _make_query_param($opts, qw/responseView/);

    return $self->request_get($path);
}

sub list_tasks {
    my ($self, $parent, $opts) = @_;
    my $path = $parent . '/tasks';

    $path .= _make_query_param($opts, qw/responseView pageSize pageToken/);

    return $self->request_get($path);
}

sub run_task {
    my ($self, $name, $opts) = @_;
    my $path = $name . ':run';

    my %param = ();
    defined $opts->{responseView} and $param{responseView} = $opts->{responseView};

    return $self->request_post($path, \%param);
}

1;


lib/Google/CloudTasks/Location.pm
package Google::CloudTasks::Location;

use Mouse;
use Google::CloudTasks::Util qw/make_query_string/;
use Google::CloudTasks::Queue;

has client => (
    is => 'ro',
    isa => 'Google::CloudTasks::Client',
    required => 1,
    weak_ref => 1,
);

has name => (
    is => 'rw',
    isa => 'Str',
);

has locationId => (
    is => 'rw',
    isa => 'Str',
);

has displayName => (
    is => 'rw',
    isa => 'Str',
);

has labels => (
    is => 'rw',
    isa => 'HashRef',
);

has metadata => (
    is => 'rw',
    isa => 'HashRef',
);

no Mouse;

__PACKAGE__->meta->make_immutable;

sub new_from_hash {
    my ($self, $client, $hash) = @_;

    my $obj = __PACKAGE__->new(
        client => $client,
    );

    for (qw/name locationId displayName labels metadata/) {
        defined $hash->{$_} and $obj->$_($hash->{$_});
    }

    return $obj;
}

sub list_queues {
    my ($self, $args) = @_;

    my $path = $self->name . '/queues';
    $path .= make_query_string($args, qw/filter pageSize pageToken/);

    my $ret = $self->client->request_get($path);
    my @queues = [
        map { Google::CloudTasks::Queue->new_from_hash($self->client, $_) }
        @{$ret->{queues}}
    ];
    return {
        queues => \@queues,
        nextPageToken => $ret->{nextPageToken},
    };
}

1;


lib/Google/CloudTasks.pm
package Google::CloudTasks;
use 5.008001;
use strict;
use warnings;
use utf8;

use Google::CloudTasks::Client;

our $VERSION = "0.01";

sub client {
    my ($class, @args) = @_;

    my %args = @args;
    if ($args{version} and $args{version} ne 'v2') {
        die "Currently only version 'v2' of CloudTasks is suppoted.";
    }
    return Google::CloudTasks::Client->new(@args);
}

1;

__END__

=encoding utf-8

=head1 NAME

Google::CloudTasks - Perl client library for the Google CloudTasks API (I<unofficial>).

=head1 SYNOPSIS

    use Google::CloudTasks;

    my $client = Google::CloudTasks->client(
        version => 'v2',
        credentials_path => '/path/to/credentials.json',
    );

    #  Create task
    my $project_id = 'myproject';
    my $location_id = 'asia-northeast1';
    my $queue_id = 'myqueue';
    my $parent = "projects/$project_id/locations/$location_id/queues/$queue_id";

    my $task = {
        name => "$parent/tasks/mytask-01234567",
        appEngineHttpRequest => {
            relativeUri => '/do_task',
        },
    }
    my $ret = $client->create_task($parent, $task);


=head1 DESCRIPTION

Google::CloudTasks L<https://cloud.google.com/tasks/docs/reference/rest/>

This is a Perl client library for the Google CloudTasks API.

=head2 AUTHENTICATION

A service account with appropriate roles is required. You need to download JSON file and specify C<credentials_path>.
See also: L<https://cloud.google.com/docs/authentication/getting-started#creating_the_service_account>

=head1 METHODS

All methods handle raw hashref (or arrayref of hashref), rather than objects.

=head2 Create a client

    my $client = Google::CloudTasks->client(
        version => 'v2',
        credentials_path => '/path/to/credentials.json',
    );

C<version> is an API version. (Currently only C<v2> is available)
C<credentials_path> is a path to a service account JSON file.

=head2 Location

Refer the detailed representation of location at L<https://cloud.google.com/tasks/docs/reference/rest/Shared.Types/ListLocationsResponse#Location>

=head3 get_location

Gets information about a location.

    my $location = $client->get_location("projects/$PROJECT_ID/locations/$LOCATION_ID");

=head3 list_locations

Lists information about all locations under project.

    my $ret = $client->list_locations("projects/$PROJECT_ID");
    my $locations = $ret->{locations};

=head2 Queue

Refer the detailed representation of queue at L<https://cloud.google.com/tasks/docs/reference/rest/v2/projects.locations.queues#Queue>

=head3 create_queue

Creates a queue.

    my $queue = {
        name => 'queue-name',
    };
    my $created = $client->create_queue("projects/$PROJECT_ID/locations/$LOCATION_ID", $queue);

=head3 delete_queue

Deletes a queue.

    $client->delete_queue("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID")

=head3 get_queue

Gets information of a queue.

    my $queue = $client->get_queue("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID");

=head3 list_queues

Lists information of all queues.

    my $ret = $client->list_queues("projects/$PROJECT_ID/locations/$LOCATION_ID");
    my $queues = $ret->{queues};

=head3 patch_queue

Updates a queue.

    my $queue = {
        retryConfig => {
            maxAttempts => 5,
        },
    };
    my $update_mask = { updateMask => 'retryConfig.maxAttempts' };
    my $updated = $client->patch_queue(
        "projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID",
        $queue,
        $update_mask,   # optional
    );

=head3 pause_queue

Pauses a queue.

    my $queue = $client->pause_queue("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID");

=head3 resume_queue

Resumes a queue.

    my $queue = $client->resume_queue("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID");

=head2 Task

Refer the detailed representation of task at L<https://cloud.google.com/tasks/docs/reference/rest/v2/projects.locations.queues.tasks#Task>

=head3 create_task

Creates a task. Note that a request body in C<appEngineHttpRequest> should be base64-encoded.

    use MIME::Base64;

    my $body = encode_base64('{"name": "TaskTest"}');
    chomp($body);

    my $task = {
        name => "projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID",
        appEngineHttpRequest => {
            relativeUri => '/path',
            headers => [
                'Content-Type' => 'application/json',
            ],
            body => $body,
        },
    };
    my $created = $client->create_task(
        "projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID",
        $task
    );

=head3 delete_task

Deletes a task.

    $client->delete_task("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID/tasks/$TASK_ID");

=head3 get_task

Gets information of a task.

    my $task = $client->get_task("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID/tasks/$TASK_ID");

=head3 list_tasks

Lists information of all tasks.

    my $ret = $client->list_tasks("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID");
    my $tasks = $ret->{tasks};

=head3 run_task

Runs a task.

    my $ret = $client->run_task("projects/$PROJECT_ID/locations/$LOCATION_ID/queues/$QUEUE_ID/tasks/$TASK_ID");

=head1 TODO

The following methods has implemented, but not tested nor documented yet.

C<Queue.getIamPolicy>

C<Queue.setIamPolicy>

C<Queue.testIamPermissions>

=head1 LICENSE

Copyright (C) egawata.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

egawata (egawa dot takashi at gmail.com)

=cut
